src/snowflake/connector/auth_usrpwdmfa.py

```diff
@@ -0,0 +1,43 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+#
+# Copyright (c) 2012-2020 Snowflake Computing Inc. All right reserved.
+#
+
+import logging
+
+from .auth_by_plugin import AuthByPlugin
+
+logger = logging.getLogger(__name__)
+
+MFA_TOKEN = 'MFATOKEN'
+
+
+class AuthByUsrPwdMfa(AuthByPlugin):
+    """Username & password & mfa authenticator."""
+
+    @property
+    def assertion_content(self):
+        return "*********"
+
+    def __init__(self, password, mfa_token=None):
+        """Initializes and instance with a password and a mfa token."""
+        self._password = password
+        self._mfa_token = mfa_token
+
+    def set_mfa_token(self, value):
+        self._mfa_token = value
+
+    def authenticate(self, authenticator, service_name, account, user, password):
+        """NOOP."""
+        pass
+
+    def update_body(self, body):
+        """Sets the password and mfa_token if available.
+
+        Don't set body['data']['AUTHENTICATOR'], since this is still snowflake default authenticator.
+        """
+        if self._password:
```

Comment:
Do you think the password might be missing at this stage? I think we should throw an exception here if the password is missing. By this stage, the password should be set already.

Reply:
Actually we have a prechecking, before this update_body function get called. And this code is only to be align with the code in auth_default.py. So basically, if no password is provided with username_password_mfa authenticator, then an exception will be thrown before this part of code is reached. Does this sound acceptable?

Reply:
update_body function call is made on the auth_instance only. There's only one auth_instance that exists at once, so if you have chosen the username_password_mfa authenticator you will only initialize the mfa auth instance and none of the checks performed in the default auth instance would apply. When you say there's already a check, do you mean check in the config parser? Can you point out where's the check you're referring to?

Reply:
Oh yeah, I referred to the check in config parser https://github.com/snowflakedb/snowflake-connector-python/blob/28682e91b15477b22ab8941a8ab4723c37fddd53/src/snowflake/connector/connection.py#L734.

Reply:
If that's the case, why's there an if clause, you can just assign it right? If you think the password will always be set then there's not need of the if condition, if you want to account for the case that it might not be set then you should raise an exception if it is not set.

Reply:
Ahh okay, because I found https://github.com/snowflakedb/snowflake-connector-python/blob/28682e91b15477b22ab8941a8ab4723c37fddd53/src/snowflake/connector/auth_default.py#L26-L29 for normal snowflake username password login, which is almost the same as username password mfa login to some extent. And if this always works, I guess not change that is better for a huge code base, since there could be some logic we are not aware of that require it as is.

Reply:
Change has to start somewhere right. Please add the exception clause if you want to be safe.

Reply:
Okay, let me add that